### PR TITLE
Fix headless UI bug

### PR DESCRIPTION
**Pull request by rimadoma**
Fixed the bug where dialog would pop open even if UI was set _headless_. Sorry it took me so long to submit this bug fix.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -126,6 +131,18 @@
...
 		</dependency>
+		<dependency>
+			<groupId>org.mockito</groupId>
+			<artifactId>mockito-core</artifactId>
+			<version>1.10.19</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.scijava</groupId>
+			<artifactId>scijava-ui-swing</artifactId>
```
Comment (ctrueden) [inline pom.xml]: This is a circular dependency: `scijava-ui-swing` depends on `scijava-common`. Could you please, rather than mocking a `SwingSDIUI`, just implement your own `UserInterface` plugin in the test class itself as a static inner class, and then use that? Then you wouldn't need Mockito either.

File: src/main/java/org/scijava/ui/DefaultUIService.java [hunk 2]
```diff
@@ -212,9 +213,9 @@ public boolean isHeadless() {
 
 	@Override
 	public UserInterface getDefaultUI() {
+		if (isHeadless()) return null;
```
Comment (ctrueden) [inline src/main/java/org/scijava/ui/DefaultUIService.java]: I really like this. It makes total sense to return `null` in a headless context. And if downstream code can't deal with it—tough. 😝

File: src/main/java/org/scijava/ui/DefaultUIService.java [hunk 3]
```diff
@@ -257,16 +258,31 @@ public UserInterface getUI(final String name) {
 
 	@Override
 	public void show(final Object o) {
-		getDefaultUI().show(o);
+		final UserInterface defaultUI = getDefaultUI();
+		if (defaultUI == null) {
```
Comment (ctrueden) [inline src/main/java/org/scijava/ui/DefaultUIService.java]: Please use the more compact one-line syntax:
```java
if (defaultUI == null) return;
```
for consistency with the other blocks of code in these here parts.

File: src/test/java/org/scijava/ui/DefaultUIServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.mockito.Mockito.mock;
+
+import org.junit.Test;
+import org.scijava.ui.swing.sdi.SwingSDIUI;
+
+/**
+ * Tests for {@link DefaultUIService}
```
Comment (ctrueden) [inline src/test/java/org/scijava/ui/DefaultUIServiceTest.java]: By convention, I have named the service tests after the interfaces. While technically maybe less correct, it produces shorter class names, and it's not like we unit test any _other_ service interface implementations.

So for consistency, let's call this `UIServiceTest` instead of `DefaultUIServiceTest`.

File: .gitignore [hunk 5]
```diff
@@ -1,5 +1,14 @@
...
+/target/
+
+# Eclipse #
 /.classpath
 /.project
 /.settings/
-/target/
+
+
+# IntelliJ #
```
Comment (ctrueden) [inline .gitignore]: Please make the gitignore changes in a separate commit.

**Comment by ctrueden** [comment]
Thank you very much, Richard. Could you please take care of my requested modifications? Then I will definitely merge it—this is a critical fix.

**Comment by ctrueden** [comment]
So, my main question is: should we, rather than returning `null` in `getDefaultUI()`, instead implement a low-priority `HeadlessUI` in SciJava Common itself?

**Comment by rimadoma** [comment]
A rather like the idea of a null object. A third option would be to return `Optional<UserInterface>` from `getDefaultUI()`, but that'd mean backwards incompatible changes to the API, which we don't want at this point. Shall I get on with the `headlessUI`? Should it be instantiated and returned from `getDefaultUI` if the object is running headless?

**Comment by ctrueden** [comment]
> Shall I get on with the `headlessUI`? Should it be instantiated and returned from `getDefaultUI` if the object is running headless?

Sounds great. Thank you for working on it!

**Comment by rimadoma** [comment]
Sorry it took so long, I was travelling. There's a lot of boiler plate code in `HeadlessUI` and its helper classes to make them null objects. Would it be OK to just return `null` values from the methods of `HeadlessUI` like in `AbstractUserInterface`?

**Comment by ctrueden** [comment]
@rimadoma Yes, I think you can simply return `null` for the main UI elements. If you run into places where downstream code assumes non-null, we can fix it. The original idea was that each UI may or may not have each of the main components—e.g., no status bar or tool bar. I don't think we need to make dummy objects for these.

**Comment by rimadoma** [comment]
Returns `null` objects now. Updated Javadoc of `UserInterface`. Commented overridden `context()` and `getContext()` in `HeadlessUI` since they work differently in the null object than specified in the `Contextual` interface.

**Comment by ctrueden** [comment]
I am in the process of cleaning this up and enhancing it a little. Will merge soon.

**Comment by ctrueden** [comment]
Whew, that took awhile. But totally worth it! Thanks again @rimadoma.
